content/docs/concepts/resources/functions.md

```diff
@@ -131,34 +131,195 @@ variables:
 
-The `version` option specifies an exact version for the provider plugin. This can be used when you need to pin to a specific version to avoid a backward-incompatible change.
+- `version`: Pass a provider plugin version that should be used when invoking the function.
 
-The `pluginDownloadURL` option gives a URL for fetching the provider plugin. It may be necessary to supply this for third-party packages (those not hosted at [https://get.pulumi.com](https://get.pulumi.com)).
+- `async`: _This option is deprecated and will be removed in a future release_
+
+### Dependencies and ordering
+
+While the direct and output forms of a provider function are equivalent in terms
+of the results they produce when invoked, they differ in how they interact with
+the rest of the Pulumi program and the order in which they may be executed.
+Specifically:
+
+- Direct form invocations execute just like any other function call in the language. Since they do not accept Pulumi `Input`s nor return Pulumi `Output`s, they are not tracked by the Pulumi engine and do not participate in the dependency graph.
+
+- Output form invocations, on the other hand, are tracked by the Pulumi engine and participate in the dependency graph. This means, for example, that Pulumi will ensure that input resources are created or updated before an invocation and that the invocation is executed before dependent resources are created or updated.
+
+In general, you should prefer the output form of a provider function when you need to ensure that the function is executed in the correct order with respect to other resources in your program. This applies particularly if you wish to simulate something like the [`dependsOn` resource option](/docs/concepts/options/dependson/), where you want to ensure that a resource is created before the function is invoked, without necessarily passing the resource as an argument to the function. While there is no `dependsOn` [invoke option](#invoke-options) for provider functions, you can achieve the same effect using Pulumi's [`all`](/docs/concepts/inputs-outputs/all/) function and `Output`'s [`apply` method](/docs/concepts/inputs-outputs/apply/):
+
+<div><pulumi-examples>
+<div><pulumi-chooser type="language" options="typescript,python,go,csharp,java,yaml"></pulumi-chooser></div>
+<div>
+<pulumi-choosable type="language" values="typescript">
+
+```typescript
+import * as gcp from "@pulumi/gcp";
+import * as pulumi from "@pulumi/pulumi";
+
+const project = new gcp.organizations.Project("project", {
+    name: "project",
+    projectId: "project",
+    billingAccount: "bills",
+    autoCreateNetwork: false,
+});
+
+// Since `project.name` is immediately available, we need to use `all` and
```

Comment:
```suggestion
// Since `project.name` isn't immediately available, we need to use `all` and
```

I think?  Or am I misreading something?  Same with the other comments.

Reply:
Yeh I don't think this is the best example. `name` and `urn` resolve at the same time for project. Ideally this would use a different resource where it's input isn't needed to make the call but it's creation is needed for the call to actually work.

Reply:
I think this is right. Because `project.name` _is_ immediately available (since it was provided as a literal), Pulumi won't block until `project` is ready. `project.urn`, however, _won't_ be available until `project` has actually been bonafide constructed, so the `all` fixes that. @Frassle is this complete nonsense?

Reply:
Ah then I misunderstood  In that case `project.name` wouldn't need to be part of `pulumi.all` at all though right?  Maybe something like 

// Since `project.name` is immediately available (before the project is ready), but we also want to wait for the urn to be ready, we need to use `all` and

Or maybe

// Since `project.name` is immediately available (before the project is ready), we can use `project.urn` as a proxy for it being ready.  This is why we need to use `all` to make sure it's completely ready

Reply:
Yea this example isn't great it turns out -- have replaced with something synthetic but at least correct (and this seems to be in line with how `dependsOn` is currently documented)